### ISSUE TITLE
Pod deletes/updates fail with "existing toleration can not be modified" error

### ISSUE THREAD

**Issue by shrinandj** [kubernetes/kubernetes#70502]
**What happened**:
We have a Kubernetes setup with a Mutating webhook enabled. The mutating webhook changes an incoming pod spec and adds tolerations to it. FWIW, the mutating webhook adds two tolerations with the same key but different values:
```
  - key: mytolerations/mykey
    value: abcd-default
  - key: mytolerations/mykey
    value: pqrs-default
```

This mutation of the incoming pod spec only happens on CREATES of the pods.

However, with this setup, we're seeing that at times pods updates and deletes are failing with the following error:
```
E1018 15: 16: 18.490361 1 garbagecollector.go: 265] error syncing item & garbagecollector.node {
	identity: garbagecollector.objectReference {
		OwnerReference: v1.OwnerReference {
			APIVersion: "v1",
			Kind: "Pod",
			Name: "api-56f4f9f448-l5nwx",
			UID: "0980d6ee-d27e-11e8-a4f5-0a8fe6bc0fca",
			Controller: ( * bool)(nil),
			BlockOwnerDeletion: ( * bool)(nil)
		},
		Namespace: "money-capital-quickbooks-capital-usw2-ppd-qal"
	},
	dependentsLock: sync.RWMutex {
		w: sync.Mutex {
			state: 0,
			sema: 0x0
		},
		writerSem: 0x0,
		readerSem: 0x0,
		readerCount: 1,
		readerWait: 0
	},
	dependents: map[ * garbagecollector.node] struct {} {},
	deletingDependents: true,
	deletingDependentsLock: sync.RWMutex {
		w: sync.Mutex {
			state: 0,
			sema: 0x0
		},
		writerSem: 0x0,
		readerSem: 0x0,
		readerCount: 0,
		readerWait: 0
	},
	beingDeleted: true,
	beingDeletedLock: sync.RWMutex {
		w: sync.Mutex {
			state: 0,
			sema: 0x0
		},
		writerSem: 0x0,
		readerSem: 0x0,
		readerCount: 0,
		readerWait: 0
	},
	virtual: false,
	virtualLock: sync.RWMutex {
		w: sync.Mutex {
			state: 0,
			sema: 0x0
		},
		writerSem: 0x0,
		readerSem: 0x0,
		readerCount: 0,
		readerWait: 0
	},
	owners: [] v1.OwnerReference {
		v1.OwnerReference {
			APIVersion: "extensions/v1beta1",
			Kind: "ReplicaSet",
			Name: "api-56f4f9f448",
			UID: "0977db9d-d27e-11e8-a4f5-0a8fe6bc0fca",
			Controller: ( * bool)(0xc423a9e84a),
			BlockOwnerDeletion: ( * bool)(0xc423a9e84b)
		}
	}
}: Pod "api-56f4f9f448-l5nwx"
is invalid: spec.tolerations: Forbidden: existing toleration can not be modified except its tolerationSeconds
```

This does not happen always.

**What you expected to happen**:
Pod deletes should succeed.

**Anything else we need to know?**:
Looking deeper into the logs, it seems that the garbagecollector in the kube-controller-manager was operating on the wrong version of the object (base on the observed `resourceVersion`)

Also, if we force delete the pod using `--force --grace-period=0`, it worked just fine.

**Environment**:
- Kubernetes version (use `kubectl version`): v1.10.7
- Cloud provider or hardware configuration: AWS
- OS (e.g. from /etc/os-release): RHEL 7
- Kernel (e.g. `uname -a`):
- Install tools: kops

Any ideas, what could be going wrong?

/cc @sandeepbhojwani

/sig api-machinery

/kind bug

**Comment by mbohlool** [comment]
cc @jennybuckley

**Comment by sandeepbhojwani** [comment]
The problem seems to be related to LimitRange. Here are the 2 scenarios..


1. Follow the steps below to create a test namespace, apply a limitrange config, create a pod, apply 2 patches to the pod and notice that the 2nd patch fails.
```
MTVL160743e61:demo-logging sbhojwani$ k create ns test
namespace/test created
MTVL160743e61:demo-logging sbhojwani$ kns test
Context "sbhojwani-1110.cluster.k8s.local" modified.
Active namespace is "test".
MTVL160743e61:demo-logging sbhojwani$ cat mem-cpu-limit.yml 
apiVersion: v1
kind: LimitRange
metadata:
  name: mem-limit-range
spec:
  limits:
  - default:
      memory: 512Mi
    defaultRequest:
      cpu: 100m
      memory: 256Mi
    type: Container
MTVL160743e61:demo-logging sbhojwani$ k apply -f mem-cpu-limit.yml 
limitrange/mem-limit-range created
MTVL160743e61:demo-logging sbhojwani$ cat pod-allow.yml
apiVersion: v1
kind: Pod
metadata:
  name: pod-allow
  labels:
    app: pod-allow
spec:
  containers:
  - name: hw1
    image: debian:9.2
    command: ['sh', '-c', 'sleep 100000']
  restartPolicy: Never
  tolerations:
  - key: k1
MTVL160743e61:demo-logging sbhojwani$ k apply -f pod-allow.yml
pod/pod-allow created
MTVL160743e61:demo-logging sbhojwani$ k patch pod pod-allow --type=json -p='[{"op":"add","path":"/spec/tolerations/-","value":{"key":"k","value":"v1"}}]'
pod/pod-allow patched
MTVL160743e61:demo-logging sbhojwani$ k patch pod pod-allow --type=json -p='[{"op":"add","path":"/spec/tolerations/-","value":{"key":"k","value":"v2"}}]'
The Pod "pod-allow" is invalid: spec.tolerations: Forbidden: existing toleration can not be modified except its tolerationSeconds
```

2. Now repeat the steps in 1. with 1 modification - dont' create LimitRange resource. Notice that both the patches work..

```
MTVL160743e61:demo-logging sbhojwani$ k create ns test4
namespace/test4 created
MTVL160743e61:demo-logging sbhojwani$ k apply -f pod-allow.yml
pod/pod-allow created
MTVL160743e61:demo-logging sbhojwani$ k patch pod pod-allow --type=json -p='[{"op":"add","path":"/spec/tolerations/-","value":{"key":"k","value":"v1"}}]'
pod/pod-allow patched
MTVL160743e61:demo-logging sbhojwani$ k patch pod pod-allow --type=json -p='[{"op":"add","path":"/spec/tolerations/-","value":{"key":"k","value":"v2"}}]'
pod/pod-allow patched
```

/cc @jennybuckley

**Comment by sandeepbhojwani** [comment]
Here's one more odd behavior. Creating a POD that already includes 2 tolerations with the same key works differently when there's a limitrange..

Without limitrange, pod gets created with 2 tolerations
```
MTVL160743e61:demo-logging sbhojwani$ cat pod-allow.yml
apiVersion: v1
kind: Pod
metadata:
  name: pod-allow
  labels:
    app: pod-allow
spec:
  containers:
  - name: hw1
    image: debian:9.2
    command: ['sh', '-c', 'sleep 100000']
  restartPolicy: Never
  tolerations:
  - key: k
    value: v1
  - key: k
    value: v2
MTVL160743e61:demo-logging sbhojwani$ k apply -f pod-allow.yml
pod/pod-allow created
MTVL160743e61:demo-logging sbhojwani$ k get po pod-allow -o json | jq .spec.tolerations
[
  {
    "key": "k",
    "value": "v1"
  },
  {
    "key": "k",
    "value": "v2"
  },
  {
    "effect": "NoExecute",
    "key": "node.kubernetes.io/not-ready",
    "operator": "Exists",
    "tolerationSeconds": 300
  },
  {
    "effect": "NoExecute",
    "key": "node.kubernetes.io/unreachable",
    "operator": "Exists",
    "tolerationSeconds": 300
  }
]

```

With limitrange, pod gets created with only the 2nd toleration. Also there's an extra toleration for `"key": "node.kubernetes.io/not-ready"`

```
MTVL160743e61:demo-logging sbhojwani$ k apply -f mem-cpu-limit.yml 
limitrange/mem-limit-range created
MTVL160743e61:demo-logging sbhojwani$ k apply -f pod-allow.yml
pod/pod-allow created
MTVL160743e61:demo-logging sbhojwani$ k get po pod-allow -o json | jq .spec.tolerations
[
  {
    "effect": "NoSchedule",
    "key": "node.kubernetes.io/memory-pressure",
    "operator": "Exists"
  },
  {
    "effect": "NoExecute",
    "key": "node.kubernetes.io/unreachable",
    "operator": "Exists",
    "tolerationSeconds": 300
  },
  {
    "key": "k",
    "value": "v2"
  },
  {
    "effect": "NoExecute",
    "key": "node.kubernetes.io/not-ready",
    "operator": "Exists",
    "tolerationSeconds": 300
  }
]

```

**Comment by sandeepbhojwani** [comment]
Root cause is a bug in these 2 functions introduced in https://github.com/kubernetes/kubernetes/commit/af5379485411214d0a1f90c7974621b0f809afb3

/cc @aveshagarwal 

```
func MergeTolerations(first []api.Toleration, second []api.Toleration) []api.Toleration {
	var mergedTolerations []api.Toleration
	mergedTolerations = append(mergedTolerations, second...)
 	firstMap := ConvertTolerationToAMap(first)
	secondMap := ConvertTolerationToAMap(second)
 	for k1, v1 := range firstMap {
		if _, ok := secondMap[k1]; !ok {
			mergedTolerations = append(mergedTolerations, v1)
		}
	}
	return mergedTolerations
}

// ConvertTolerationToAMap converts toleration list into a map[string]api.Toleration
func ConvertTolerationToAMap(in []api.Toleration) map[key]api.Toleration {
	out := map[key]api.Toleration{}
	for i := range in {
		out[key{in[i].Key, in[i].Effect}] = in[i]
	}
	return out
}
```
The ConvertTolerationToAMap function uses toleration's key and effect only to generate the map key. So if i have 2 tolerations(as in my example in earlier comment) with the same key and effect but different values then the ConvertTolerationToAMap will override the first toleration with the second toleration.

**Comment by liggitt** [comment]
cc @jpbetz for impact on admission reinvocation

**Comment by jpbetz** [comment]
Thanks @liggitt.

Right now our toleration merge logic is to treat any two tolerations with the same (key, effect) as merge-able. But after reading through the taints and tolerations documentation I'm not sure that's right. `tolerationSeconds` appears to be the only field that does not somehow identify the toleration w/r/t how it matches against taints. I'm curious what if there would be any ill effects to changing the merge logic to treat only two tolerations with the same (key, value, effect, operator) as merge-able?

**Comment by liggitt** [comment]
at the very least no updates should be made by the toleration admission plugin on update that would break pod validation. I'm unsure what the desired behavior is on create

**Comment by makoscafee** [comment]
@aveshagarwal Hello! I'm bug triage shadow for the 1.16 release cycle and considering this issue is tagged for 1.16, but not updated for a long time, I'd like to check its status. The code freeze is starting on August 29th (about 1.5 weeks from now), which means that there should be a PR ready (and merged) until then.

Do we still target this issue to be fixed for 1.16?

**Comment by liggitt** [comment]
fixed by https://github.com/kubernetes/kubernetes/pull/81732

/close
